### ISSUE TITLE
Lambda function parameters leaked to outside environment

### ISSUE THREAD

**Issue by SebastianS90** [Whiley/WhileyCompiler#612]
Shouldn't parameters from lambda's only be valid in the corresponding expression?

````
type myfun is function(int)->(int)

public export method test():
	myfun f = &(int x -> x)
	int x = f(1)
	assume x == 1
````

`internal failure, Variable already declared - x` reported at `int x = f(1)`

Let's try even more evil stuff:

````
type myfun is function(int)->(int)

public export method test():
	myfun f = &(int x -> x)
	x = 1
````
Gives an internal failure caused by `NullPointerException` in `wyc.builder.CodeGenerator.generateAssignment`
At least we cannot write to arbitrary memory :+1:

**Comment by DavePearce** [comment]
Yes, that looks like a bug :)

**Comment by SebastianS90** [comment]
fixed by #617
